existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,194 @@
+In order to use them ... you still need to use `forall.` keyword ¯\\\_(ツ)_/¯.
+That might be a little confusing.
+The following construction is called existential type in Haskell - when you use a forall quantifier in the right-hand side of function or data type declaration.
+But what is it used for? Let's find out!
+
+### Usage
+
+We will start with functions:
+
+```haskell
+func :: forall a c. a -> (forall b. (Show b) => b) -> c
+```
+Few things to mention.
+Here `a` and `c` are universally quantified and `b` is existentially quantified variables.
+The same variable can't be quantified more than once.
+Also, you can see that constraints for `b` are added in the middle of the function type declaration.
+This example is pretty useless.
+The only thing that has changed here is that now you can't manually choose type for `b` - the compiler must deduce the correct variable you want to use by itself.
+
+When we come to data type declaration, it becomes more useful.
+
+#### Hiding type variables
+```Haskell
+data MyData a = forall b. MkMyData b a
+```
+The first thing you might have noticed is that we don't have a type variable for `b` on the left side of expression.
+However, we still can use it on the right side.
+This is because `b` is existentially quantified.
+
+```Haskell
+MkMyData @_ @Integer -- This won't work, you don't have second variable in declaration
+```
+Compiler will deduce the type of the variable itself, and you can't specify it.
+
+That is actually why this construction is called existential type - because you tell a compiler, that there exists type, that
+can be used in constructor, without specifying which one exactly. Also it is called so, because from type theory's point of view
+this constructions are isomorphic:
+
+```haskell
+data MyTypeWithFunc = forall a. MkMyTypeWithFunc a (a -> String)
+
+MkMyTypeWithFunc :: forall a . a -> (a -> String)-> MyTypeWithFunc
+MkMyTypeWithFunc :: (exists a . (a, a -> String)) -> MyTypeWithFunc
+```
+
+#### Heterogeneous lists
+
+Now we understand that existential types allow you to introduce variables, without necessity to specify them.
+This might be useful when building a multi-typed list:
+
+```haskell
+data Elem = forall a. Elem a
+
+multitypedList :: [Elem]
+multitypedList = [Elem "a", Elem 1, Elem (Just 5)]
+```
+This is a heterogeneous list that contains variables of different types.
+You couldn't do that without existential types, because even if you use type variables, you should still have every element in the list have the same type.
+
+Note, that since type variable `a` is hidden with quantifier, you can't add constraints to it, beside those added to the constructor
+
+```haskell
+showFunc :: Show ??? => [Elem] -> String -- There is no type variable, that you can add constraint to
+showFunc (x:xs) => (show x) ++ (showFunc xs)
+```
+
+The same "shadowing" is used in well-known `SomeException` wrapper:
```

Comment:
What do you mean by "shadowing" in this context?

Reply:
By shadowing I meant hiding types of fields. Maybe it has a more appropriate term?